### PR TITLE
Merge the new UnitTesting solution crawler into Main

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@tmat Please review.  This is merging all the solution-crawler work done in the feature branch back into main :)

**Review by jmarolf — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@tmat review please :)

File: src/Workspaces/Core/Portable/SolutionCrawler/InvocationReasons.cs [hunk 1]
```diff
@@ -2,44 +2,39 @@
...
         public InvocationReasons(string reason)
-            : this(ImmutableHashSet.Create<string>(reason))
+            : this(ImmutableHashSet.Create(reason))
         {
         }
 
-        private InvocationReasons(ImmutableHashSet<string> reasons)
-            => _reasons = reasons;
+        public InvocationReasons(ImmutableHashSet<string> reasons)
+            => _reasons = reasons ?? ImmutableHashSet<string>.Empty;
```
Comment (tmat) [inline src/Workspaces/Core/Portable/SolutionCrawler/InvocationReasons.cs:26]: >reasons [](http://example.com/codeflow?start=26&length=7)

Can this actually be null? There is no `?` in the signature.

File: src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/Api/UnitTestingInvocationReasonsWrapper.cs [hunk 2]
```diff
@@ -19,8 +19,8 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.Api
         public static readonly UnitTestingInvocationReasonsWrapper PredefinedSyntaxChanged = new(PredefinedInvocationReasons.SyntaxChanged);
         public static readonly UnitTestingInvocationReasonsWrapper PredefinedProjectConfigurationChanged = new(PredefinedInvocationReasons.ProjectConfigurationChanged);
         public static readonly UnitTestingInvocationReasonsWrapper PredefinedDocumentOpened = new(PredefinedInvocationReasons.DocumentOpened);
-        public static readonly UnitTestingInvocationReasonsWrapper PredefinedDocumentRemoved = new(PredefinedInvocationReasons.DocumentRemoved);
         public static readonly UnitTestingInvocationReasonsWrapper PredefinedDocumentClosed = new(PredefinedInvocationReasons.DocumentClosed);
```
Comment (tmat) [inline src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/Api/UnitTestingInvocationReasonsWrapper.cs:22]: >readonly [](http://example.com/codeflow?start=22&length=8)

Nit: no change here, just reodering

File: src/Features/Core/Portable/LegacySolutionEvents/ILegacySolutionEventsAggregationService.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    /// This is a legacy api intended only for existing SolutionCrawler partners to continue to function (albeit with
+    /// ownership of that crawling task now belonging to the partner team, not roslyn).  It should not be used for any
+    /// new services.
+    /// </summary>
+    internal interface ILegacySolutionEventsAggregationService : IWorkspaceService
+    {
+        bool ShouldReportChanges(SolutionServices services);
+
+        ValueTask OnWorkspaceChangedAsync(WorkspaceChangeEventArgs args, CancellationToken cancellationToken);
+#if false // Not used in unit testing crawling
```
Comment (tmat) [inline src/Features/Core/Portable/LegacySolutionEvents/ILegacySolutionEventsAggregationService.cs:27]: >false [](http://example.com/codeflow?start=4&length=5)

Why keep this?

File: src/Features/Core/Portable/ExternalAccess/UnitTesting/API/INewUnitTestingIncrementalAnalyzerImplementation.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.SolutionCrawler;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting.Api
+{
+    internal interface INewUnitTestingIncrementalAnalyzerImplementation
+    {
+#if false // Not used in unit testing crawling
```
Comment (tmat) [inline src/Features/Core/Portable/ExternalAccess/UnitTesting/API/INewUnitTestingIncrementalAnalyzerImplementation.cs:13]: >false [](http://example.com/codeflow?start=4&length=5)

Why keep this?

**Review by tmat — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@shyamnamboodiripad  this is now merged.  PTAL when you can and see about moving onto this :)  LMK if you run into any issues.  Thanks!

**Pull request merged**